pkg/cri/opts/spec_linux.go

```diff
@@ -139,19 +140,26 @@ func IsCgroup2UnifiedMode() bool {
 }
 
 // WithCDI updates OCI spec with CDI content
-func WithCDI(annotations map[string]string) oci.SpecOpts {
+func WithCDI(annotations map[string]string, CDIDevices []*runtime.CDIDevice) oci.SpecOpts {
 	return func(ctx context.Context, _ oci.Client, c *containers.Container, s *oci.Spec) error {
-		// TODO: Once CRI is extended with native CDI support this will need to be updated...
-		_, cdiDevices, err := cdi.ParseAnnotations(annotations)
-		if err != nil {
-			return fmt.Errorf("failed to parse CDI device annotations: %w", err)
-		}
-		if cdiDevices == nil {
-			return nil
+		var cdiDevices []string
+		var err error
+		if len(CDIDevices) > 0 {
+			for _, CDIDevice := range CDIDevices {
+				cdiDevices = append(cdiDevices, CDIDevice.Name)
+			}
+			log.G(ctx).Infof("container %v: CDI devices from CRI Config.CDIDevices: %v", c.ID, cdiDevices)
+		} else {
```

Comment:
is it a good idea to ignore the annotations if CDIDevices are passed via CRI is it possible one might be missed by kubelet or another CRI api user?  I've not checked the kubelet code. If possible to use both... perhaps we should add both to the list with a warning

Reply:
either way I think we need a "first" debug out that this annotation use is deprecated see the container config option in v1 cri with commit #..

and a TODO: here notifying that the annotations are deprecated

Reply:
I'd prefer to only use annotations or CRI field to avoid complicating the implementation. Kubelet sets both currently, but I'm going to get rid of annotations usage there as soon as Containerd and CRIO support CRI field.

Added Debug output and TODO as suggested.

Reply:
speaking to "kubelet sets both currently".. does kubelet restrict/filter annotations of this type set by users in the pod spec? 

so long as we error out I think "or" is a fair decision... otherwise how would the user know one device annotation or one cri field device annotation is being ignored...

Reply:
@mikebrow
> does kubelet restrict/filter annotations of this type set by users in the pod spec?

It doesn't.

After discussing this topic with CDI folks we agreed to support both CDIDevices **and** annotations until annotations are deprecated. In addition to the reasons you've mentioned above, some device plugins set CDI annotations, so it makes sense to consider both CRI CDIDevices field and annotations on the runtime side.

I've updated the PR and the tests accordingly, PTAL.